### ISSUE TITLE
JMX implementation : feature parity for target systems

### ISSUE THREAD

**Issue by SylvainJuge** [open-telemetry/opentelemetry-java-instrumentation#12158]
JMX Insights supports some values for `otel.jmx.target.system`, those are defined in YAML files [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/jmx-metrics/javaagent/src/main/resources/jmx/rules).

JMX Gatherer (in contrib) supports more values of `otel.jmx.target.system`, those are defined in Groovy scripts [here](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/jmx-metrics/src/main/resources/target-systems).

While the Groovy scripts are convenient, moving to YAML seems a more future-proof solution:
- removes security risk of having executable groovy scripts
- YAML syntax is already widespread and usually do not require java/groovy knowledge
- YAML syntax could later allow to inline the configuration in a global OpenTelemetry YAML configuration when such would be available, for now it has to be stored in a separate file.

Merging both implementations and bringing them to feature parity means that we have to attempt migrate/align all of the JMX Gatherer supported systems and ensure they can be implemented with YAML. Doing so will highlight any missing feature of the YAML implementation by adding any missing part.

Once the alignment is complete, we should then be able to start on the next step: building a "JMX Scraper" in contrib based on the YAML implementation in instrumentation.

For each system listed below, we need to ensure the following with JMX Insights
- add YAML if system is not supported yet
- convert groovy metrics to their YAML equivalent
- deal with any found inconsistency for existing metrics by choosing to
    - leave them as-is
    - fix YAML or Groovy definitions (or both)
- add any missing feature to YAML implementation if needed

List of systems to cover:

- [ ] activemq [activemq.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/activemq.groovy)
- [ ] cassandra [cassandra.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/cassandra.groovy)
- [ ] hadoop [hadoop.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/hadoop.groovy)
- [ ] hbase [hbase.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/hbase.groovy)
- [ ] jetty [jetty.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/jetty.groovy)
- [ ] jvm [jvm.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/jvm.groovy)
- [ ] kafka [kafka.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/kafka.groovy)
- [ ] kafka-consumer [kafka-consumer.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/kafka-consumer.groovy)
- [ ] kafka-producer [kafka-producer.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/kafka-producer.groovy)
- [ ] solr [solr.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/solr.groovy)
- [ ] tomcat [tomcat.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/tomcat.groovy)
- [ ] wildfly [wildfly.groovy](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/wildfly.groovy)

**Comment by SylvainJuge** [comment]
Ping @robsunday I can't yet co-assign you as you are not part of the otel contributors group.

**Comment by robsunday** [comment]
I'll look on Jetty

**Comment by SylvainJuge** [comment]
For Wildfly, the mapping is also not the same but equivalent, there isn't anything we need to add for 1:1 support beyond aligning the metrics themselves.

- JMX: `jboss.as:deployment=*,subsystem=undertow`
  - Both map `deployment` => `deployment` attribute
  - `sessionsCreated`: `wildfly.session.count` <==> `wildfly.session.sessionsCreated`
  - `activeSessions`: `wildfly.session.active` <==> `wildfly.session.activeSessions`
  - `expiredSessions`: `wildfly.session.expired` <==> `wildfly.session.expiredSessions`
  - `rejectedSessions`: `wildfly.session.rejected` <==> `wildfly.session.rejectedSessions`
- JMX: `jboss.as:subsystem=undertow,server=*,http-listener=*`
  - Both map `server` => `server` attribute and `http-listener` => value of `listener`
  - `requestCount`:  `wildfly.request.count` <==> `wildfly.request.requestCount`
  - `processingTime`: `wildfly.request.time` <==> `wildfly.request.processingTime`
  - `errorCount`: `wildfly.request.server_error` <==> `wildfly.request.errorCount`
  - `bytesSent`: `wildfly.network.io` with extra `state` = `out` attribute <==> same
  - `bytesReceived`: `wildfly.network.io` with extra `state` = `in` attribute <==> same
- JMX: `jboss.as:subsystem=datasources,data-source=*,statistics=pool`
  - Both map `data-source` => value of `data_source`
  - `ActiveCount` : `wildfly.jdbc.connection.open` with `state` = `active` <==> `wildfly.db.client.connections.usage` with `state` = `used`
  - `IdleCount` :  `wildfly.jdbc.connection.open` with `state` = `idle` <==> `wildfly.db.client.connections.usage` with `state` = `idle`
  - `WaitCount`: `wildfly.jdbc.request.wait` <==> `wildfly.db.client.connections.WaitCount`
- JMX: `jboss.as:subsystem=transactions`
  - `numberOfTransactions`: `wildfly.jdbc.transaction.count` <==> `wildfly.db.client.transaction.NumberOfTransactions` 
  - `numberOfSystemRollbacks`: `wildfly.jdbc.rollback.count` with `cause` = `system` <==> `wildfly.db.client.rollback.count` with `cause` = `system`
  - `numberOfResourceRollbacks`: `wildfly.jdbc.rollback.count` with `cause` = `resource` <==> `wildfly.db.client.rollback.count` with `cause` = `resource`
  - `numberOfApplicationRollbacks`: `wildfly.jdbc.rollback.count` with `cause` = `application` <==> `wildfly.db.client.rollback.count` with `cause` = `application`

**Comment by robsunday** [comment]
Here are my findings regarding `jetty`:
- JMX: `org.eclipse.jetty.server.session:context=*,type=sessionhandler,id=*`
  - MBean property: `sessionsCreated` --> YAML: `jetty.session.sessionsCreated` <==> Groovy: `jetty.session.count`
  - MBean property: `sessionTimeTotal` --> YAML: `jetty.session.sessionTimeTotal` <==> Groovy: `jetty.session.time.total`
    - minor difference in type: YAML: `counter` / Groovy: `UpDownCounter` 
  - MBean property: `sessionTimeMax` --> YAML: `jetty.session.sessionTimeMax` <==> Groovy: `jetty.session.time.max`
  - MBean property: `sessionTimeMean` --> YAML: `jetty.session.sessionTimeMean`, not used in Groovy

- JMX: `org.eclipse.jetty.util.thread:type=queuedthreadpool,id=*` 
  - MBean property: `busyThreads` --> YAML: `jetty.threads.busyThreads` <==> Groovy: `jetty.thread.count` with extra `state=busy` attribute
      - minor difference in type: YAML: `updowncounter` / Groovy: `Value`
  - MBean property: `idleThreads` --> YAML: `jetty.threads.idleThreads` <==> Groovy: `jetty.thread.count` with extra `state=idle` attribute
      - minor difference in type: YAML: `updowncounter` / Groovy: `Value`
  - MBean property: `maxThreads` --> YAML: `jetty.threads.maxThreads`, not used in Groovy
  - MBean property: `queueSize` --> YAML: `jetty.threads.queueSize` <==> Groovy: `jetty.thread.queue.count`
      - minor difference in type: YAML: `updowncounter` / Groovy: `Value`

- JMX: `org.eclipse.jetty.io:context=*,type=managedselector,id=*`
    - MBean property: `selectCount` --> YAML: `jetty.io.selectCount` <==> Groovy: `jetty.select.count`
      - difference in units: YAML: `1` / Groovy: `{operations}`

- JMX: `org.eclipse.jetty.logging:type=jettyloggerfactory,id=*` not used in Groovy

**Comment by SylvainJuge** [comment]
For `hbase`, there isn't anything in JMX Insight for it, the mappings are simple and it should be quite straightforward (but a bit tedious) to produce an equivalent YAML to [`hbase.groovy`](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/jmx-metrics/src/main/resources/target-systems/hbase.groovy).

**Comment by SylvainJuge** [comment]
For `hadoop`:

JMX attribute `tag.Hostname` is always mapped to `node_name` metric attribute in both implementations.

JMX `Hadoop:service=NameNode,name=FSNamesystem`:
- `CapacityUsed` : `hadoop.name_node.capacity.usage` <==> `hadoop.capacity.CapacityUsed`
- `CapacityTotal`: `hadoop.name_node.capacity.limit` <==> `hadoop.capacity.CapacityTotal`
- `BlocksTotal`: `hadoop.name_node.block.count` <==> `hadoop.block.BlocksTotal`
- `MissingBlocks`: `hadoop.name_node.block.missing` <==> `hadoop.block.MissingBlocks`
- `CorruptBlocks`: `hadoop.name_node.block.corrupt` <==> `hadoop.block.CorruptBlocks`
- `VolumeFailuresTotal`: `hadoop.name_node.volume.failed` <==> `hadoop.volume.VolumeFailuresTotal`
- `FilesTotal`: `hadoop.name_node.file.count` <==> `hadoop.file.FilesTotal`
- `TotalLoad`: `hadoop.name_node.file.load` <==> `hadoop.file.TotalLoad`
- `NumLiveDataNodes`: `hadoop.name_node.data_node.count` with `state` = `live` <==> `hadoop.datenode.Count`, same `state` value (yes, there is a typo in `datanode`)
- `NumDeadDataNodes`: `hadoop.name_node.data_node.count` with `state` = `dead` <==> `hadoop.hadoop.datenode.Count`, same `state` value

**Comment by SylvainJuge** [comment]
For `cassandra`:

There is no mapping in YAML, the mapping is verbose and the lack of support for templates or string interpolation would make it quite tedious to write, but it's more an annoyance than a really blocking issue.

For example, few examples of MBeans:
- `org.apache.cassandra.metrics:type=ClientRequest`
- `org.apache.cassandra.metrics:type=ClientRequest,scope=RangeSlice`
- `org.apache.cassandra.metrics:type=ClientRequest,scope=Read`
- `org.apache.cassandra.metrics:type=ClientRequest,scope=Write`
- all of above with `scope=` with 3 variants by adding `,name=` with value in `Unavailables`, `Timeouts` or `Failures`
- `org.apache.cassandra.metrics:type=Storage,name=Load`

There isn't anything that could not be mapped using YAML syntax.

**Comment by robsunday** [comment]
For `activemq` everything except property descriptions seems to be in sync:
- JMX: `org.apache.activemq:type=Broker,brokerName=*,destinationType=Queue,destinationName=*` and `org.apache.activemq:type=Broker,brokerName=*,destinationType=Topic,destinationName=*`
  - `ProducerCount`: `activemq.producer.count` <==> `activemq.ProducerCount`
  - `ConsumerCount`: `activemq.consumer.count` <==> `activemq.ConsumerCount`
  - `MemoryPercentUsage`: `activemq.memory.usage` <==> `activemq.memory.MemoryPercentUsage`
  - `QueueSize`: `activemq.message.current` <==> `activemq.message.QueueSize`
  - `ExpiredCount`: `activemq.message.expired` <==> `activemq.message.ExpiredCount`
  - `EnqueueCount`: `activemq.message.enqueued` <==> `activemq.message.EnqueueCount`
  - `DequeueCount`: `activemq.message.dequeued` <==> `activemq.message.DequeueCount`
  - `AverageEnqueueTime`: `activemq.message.wait_time.avg` <==> `activemq.message.AverageEnqueueTime`

All `desc` fields in properties needs to be synchronized because wording is different

- JMX: `org.apache.activemq:type=Broker,brokerName=*`
  - `CurrentConnectionsCount`: `activemq.connection.count` <==> `activemq.connections.CurrentConnectionsCount`
  - `StorePercentUsage`: `activemq.disk.store_usage` <==> `activemq.disc.StorePercentUsage`
  - `TempPercentUsage`: `activemq.disk.temp_usage` <==> `activemq.disc.TempPercentUsage`

**Comment by robsunday** [comment]
`solr` case is very similar to `hbase`. No YAML at the moment but creating it should not be an issue.

**Comment by SylvainJuge** [comment]
For `kafka`, the YAML is [`kafka-broker.yaml`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/src/main/resources/jmx/rules/kafka-broker.yaml)

JMX: `kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec`:
- `Count` : `kafka.message.count`
JMX: `kafka.server:type=BrokerTopicMetrics,name=TotalProduceRequestsPerSec`:
- `Count`: `kafka.request.count` with `type` = `produce`
JMX: `kafka.server:type=BrokerTopicMetrics,name=TotalFetchRequestsPerSec`:
- `Count`: `kafka.request.count` with `type` = `fetch`
JMX: `kafka.server:type=BrokerTopicMetrics,name=FailedProduceRequestsPerSec`:
- `Count`: `kafka.request.failed` with `type` = `produce`
JMX: `kafka.server:type=BrokerTopicMetrics,name=FailedFetchRequestsPerSec`:
- `Count`: `kafka.request.failed` with `type` = `fetch`

I haven't checked in detail all the others, but they look identical between the two implementations.

I discovered that we have a way to use multiple mbeans names with the same metrics definition as seen in [kafka-broker.yaml](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/jmx-metrics/javaagent/src/main/resources/jmx/rules/kafka-broker.yaml#L54)

For `kafka-consumer.groovy` and `kafka-producer.groovy` there is no equivalent YAML mapping though.

**Comment by SylvainJuge** [comment]
Alignment has been done for `jvm`, `tomcat` and `jetty` are already merged, `wildfly` is should be merged soon (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/14208)

I have created sub-issues (one per system) to further split ongoing work on this for more clarity.

**Comment by SylvainJuge** [comment]
This issue focuses on the migration of systems supported JMX Gatherer that were not supported or supported with differences in JMX Insights and JMX Scraper, however there are also some systems that were supported in the latter and not in JMX Gatherer.

This is the case of Apache Camel, for which @robsunday opened https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/16088 to enhance it
